src/ocean/text/json/JsonDeserializer.d

```diff
@@ -0,0 +1,609 @@
+
+    test!("==")(data, TestData(["Test1", "Test2"]));
+}
+
+/// It should parse a nested struct
+unittest
+{
+    mixin(Typedef!(int, "Other", 41));
+
+    struct NestedStruct
+    {
+        ulong price;
+        Other other;
+    }
+
+    struct TestStruct
+    {
+        int id;
+        string name;
+
+        NestedStruct[] items;
+    }
+
+    string serializedTestData = `{
+        "id": 1,
+        "name": "didibao",
+        "items": [{
+            "price": 777,
+            "other": 1999
+        }]
+    }`;
+
+    auto parser = new JsonDeserializer();
+    auto result = parser.fromJsonString!(TestStruct)(serializedTestData);
+
+    TestStruct expected;
+    expected.id = 1;
+    expected.name = "didibao";
+    expected.items ~= NestedStruct(777, Other(1999));
+
+    test!("==")(result, expected);
+}
+
+version (UnitTest)
+{
+    mixin(Typedef!(int, "BufferItem", 41));
+
+    struct Buffer
+    {
+        BufferItem[] elements;
+
+        void push (BufferItem value)
+        {
+            elements ~= value;
+        }
+    }
+
+    struct TestStruct
+    {
+        Buffer buffer;
+    }
+}
+
+/// It should parse a struct with push method and an elements array
+unittest
+{
+    string serializedTestData = `{"buffer":[1,2,3]}`;
+
+    auto parser = new JsonDeserializer();
+    auto result = parser.fromJsonString!(TestStruct)(serializedTestData);
+
+    TestStruct expected;
+
+    expected.buffer.push(BufferItem(1));
+    expected.buffer.push(BufferItem(2));
+    expected.buffer.push(BufferItem(3));
+
+    test!("==")(result, expected);
+}
+
+/******************************************************************************
+
+    Generate D code that iterates over the struct member that can be
+    deserialized and calls `deserializeJson` for them.
+
+    Properties:
+        T = The struct type that is needed to be deserialized
+
+    Returns:
+        The D code that calls `deserializeJson` for each member
+
+******************************************************************************/
+
+private string genAttributeReader ( T ) ( )
+{
+    string result;
+
+    foreach ( member; properties!(T) )
+    {
+        result ~=
+            `if(member == "` ~ member ~ `")
+            {
+                result.` ~ member ~ ` = deserializeJson!(typeof(T.` ~ member ~ `))(value);
+            }`;
+    }
+
+    return result;
+}
+
+/******************************************************************************
+
+    Get the list of struct members that can be deserialized from Json
+
+    Properties:
+        T = The struct that will be checked
+
+    Returns:
+        A list of strings containing the name of the deserializable members
+
+******************************************************************************/
+
+private string[] properties ( T ) ( )
+{
+    string[] names;
+
+    foreach ( name; [ __traits(allMembers, T) ] )
+    {
+        if ( isProperty!(T)(name) )
+        {
+            names ~= name;
```

Comment:
So this is going to allocate every time it is called right?

Reply:
yes, but it's used only at ctfe, in `genAttributeReader` and it will not allocate at runtime